plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp

```diff
@@ -16,691 +19,327 @@ extern const char* const state_history_plugin_abi;
-                 "\"block_num\":${this_block}, \"block_id\":${this_id}}} to send queue",
-                 ("head", result.head.block_num)("last_irr", result.last_irreversible.block_num)
-                 ("this_block", result.this_block ? result.this_block->block_num : fc::variant())
-                 ("this_id", result.this_block ? fc::variant{result.this_block->block_id} : fc::variant{}));
-      }
+   get_status_result_v1 fill_current_status_result() {
+      get_status_result_v1 ret;
 
-      --request.max_messages_in_flight;
-      need_to_send_update = to_send_block_num <= current &&
-                            to_send_block_num < request.end_block_num;
+      ret.head              = {controller.head_block_num(), controller.head_block_id()};
+      ret.last_irreversible = {controller.last_irreversible_block_num(), controller.last_irreversible_block_id()};
+      ret.chain_id          = controller.get_chain_id();
+      if(trace_log)
+         std::tie(ret.trace_begin_block, ret.trace_end_block) = trace_log->block_range();
+      if(chain_state_log)
+         std::tie(ret.chain_state_begin_block, ret.chain_state_end_block) = chain_state_log->block_range();
+      if(finality_data_log)
+         std::tie(ret.finality_data_begin_block, ret.finality_data_end_block) = finality_data_log->block_range();
 
-      std::make_shared<blocks_result_send_queue_entry<session>>(this->shared_from_this(), std::move(result))->send_entry();
+      return ret;
    }
 
-   bool no_request_or_not_max_messages_in_flight() {
-      if (!current_request)
-         return true;
+   boost::asio::awaitable<void> write_log_entry(std::optional<locked_decompress_stream>& log_stream, std::optional<state_history_log>& log, chain::block_num_type block_num) {
+      uint64_t unpacked_size = 0;
 
-      uint32_t max_messages_in_flight = std::visit(
-         [&](auto& request) -> uint32_t { return request.max_messages_in_flight; },
-         *current_request);
+      if(log_stream) //will be unset if either request did not ask for this log entry, or the log isn't enabled
+         unpacked_size = log->get_unpacked_entry(block_num, *log_stream); //will return 0 if log does not include the block num asked for
 
-      return !max_messages_in_flight;
-   }
+      if(unpacked_size) {
+         char buff[1024*1024];
+         fc::datastream<char*> ds(buff, sizeof(buff));
+         fc::raw::pack(ds, true);
+         history_pack_varuint64(ds, unpacked_size);
+         co_await stream.async_write_some(false, boost::asio::buffer(buff, ds.tellp()));
 
-   void send_update(const state_history::block_position& head, const chain::signed_block_ptr& block, const chain::block_id_type& id) {
-      if (no_request_or_not_max_messages_in_flight()) {
-         session_mgr.pop_entry(false);
-         return;
+         ///TODO: why is there an uncompressed option in the variant?! Shouldn't it always be compressed? was this for old unit tests?
```

Comment:
This actually goes away when taking the log refactor in the next PR (https://github.com/AntelopeIO/spring/pull/237/files#diff-49a9fe1e5824be2e758a6fa48b12a86d7922882aa15a5ef54659df3543e2e8c9). While I suggested that reviewing this PR prior to #237 would still cohesively make sense stand alone, I suppose this is one of the areas where it doesn't. 

The SHIP log always contains a zlib stream yet the `locked_decompress_stream` has a variant which is ether a zlib stream or an uncompressed buffer. That uncompressed buffer is used solely for (the removed) unit tests where substantial parts of many pieces of ship are mocked. It can never be encountered during nodeos operation. The comment was added early on where the encountering the presence of anything other than a zlib stream was at odds with how ship was expected to work.